substratevm/src/com.oracle.svm.core.windows/src/com/oracle/svm/core/windows/WindowsSystemPropertiesSupport.java

```diff
@@ -122,13 +132,194 @@ private static String toJavaString(WCharPointer wcString, UnsignedWord length) {
         return CTypeConversion.toJavaString((CCharPointer) wcString, SizeOf.unsigned(WCharPointer.class).multiply(length), StandardCharsets.UTF_16LE);
     }
 
+    private Pair<String, String> cachedOsNameAndVersion;
+
+    @Override
+    protected String osNameValue() {
+        if (cachedOsNameAndVersion == null) {
+            cachedOsNameAndVersion = getOsNameAndVersion();
+        }
+        return cachedOsNameAndVersion.getLeft();
+    }
+
     @Override
     protected String osVersionValue() {
-        ByteBuffer versionBytes = ByteBuffer.allocate(4);
-        versionBytes.putInt(SysinfoAPI.GetVersion());
-        int majorVersion = versionBytes.get(3);
-        int minorVersion = versionBytes.get(2);
-        return majorVersion + "." + minorVersion;
+        if (cachedOsNameAndVersion == null) {
+            cachedOsNameAndVersion = getOsNameAndVersion();
+        }
+        return cachedOsNameAndVersion.getRight();
+    }
+
+    public Pair<String, String> getOsNameAndVersion() {
+        /*
+         * Reimplementation of code from java_props_md.c
+         */
+        SysinfoAPI.OSVERSIONINFOEXA ver = StackValue.get(SysinfoAPI.OSVERSIONINFOEXA.class);
+        ver.dwOSVersionInfoSize(SizeOf.get(SysinfoAPI.OSVERSIONINFOEXA.class));
+        SysinfoAPI.GetVersionExA(ver);
+
+        boolean is64bit = true; /* ATM we only support 64-bit Windows OS's */
```

Comment:
I implemented a workaround similar to [this](https://stackoverflow.com/a/5940770), since it was the least painful fix to do.

In the improbable case that in the future there are supported Windows systems which are using 64 bit architecture which name doesn't end with "64", this class will require changes anyway in order to support that system's naming convention.